src/environmentd/tests/server.rs

```diff
@@ -132,64 +132,238 @@ fn test_http_sql() -> Result<(), Box<dyn Error>> {
+        },
+        TestCase {
+            query: "select $1+$2 as col",
+            params: Some(vec![
+                ("1", "pg_catalog.int4"),
+                ("2", "materialize.pg_catalog.int8"),
+            ]),
+            status: StatusCode::OK,
+            body: r#"{"results":[{"rows":[[3]],"col_names":["col"]}]}"#,
+        },
+        // Parameters can be present and empty
+        TestCase {
+            query: "select 3 as col",
+            params: Some(vec![]),
+            status: StatusCode::OK,
+            body: r#"{"results":[{"rows":[[3]],"col_names":["col"]}]}"#,
+        },
+        // Ininity, NaN
+        TestCase {
+            query: "select $1+$2 as col",
+            params: Some(vec![("infinity", "numeric"), ("2", "int8")]),
+            status: StatusCode::OK,
+            body: r#"{"results":[{"error":"invalid input syntax for type numeric: \"infinity\""}]}"#,
+        },
+        TestCase {
+            query: "select $1+$2 as col",
+            params: Some(vec![("nan", "numeric"), ("2", "int8")]),
+            status: StatusCode::OK,
+            body: r#"{"results":[{"rows":[["NaN"]],"col_names":["col"]}]}"#,
+        },
+        // Quotes escaped
+        TestCase {
+            query: "select length($1), length($2)",
+            params: Some(vec![("abc", "text"), ("'abc'", "text")]),
+            status: StatusCode::OK,
+            body: r#"{"results":[{"rows":[[3,5]],"col_names":["length","length"]}]}"#,
+        },
+        // Various forms of SQL injection fail
+        TestCase {
+            query: "select length($1)",
+            params: Some(vec![("'abc'); DELETE FROM users; SELECT version(", "text")]),
+            status: StatusCode::OK,
+            body: r#"{"results":[{"rows":[[42]],"col_names":["length"]}]}"#,
+        },
+        TestCase {
+            query: "select length($1)",
+            params: Some(vec![(
+                "abc' AS text); DELETE FROM users; SELECT CAST('abc",
+                "text",
+            )]),
+            status: StatusCode::OK,
+            body: r#"{"results":[{"rows":[[50]],"col_names":["length"]}]}"#,
+        },
+        TestCase {
+            query: "select length($1)",
+            params: Some(vec![("1", "text); DELETE * FROM t; SELECT version(")]),
+            status: StatusCode::BAD_REQUEST,
+            body: r#"invalid parameter: CAST ('<val>' as <type>) must be valid expression"#,
+        },
+        // All parameters values treated as strings
+        TestCase {
+            query: "select length($1), length($2)",
+            params: Some(vec![("sum(a)", "text"), ("SELECT * FROM t;", "text")]),
+            status: StatusCode::OK,
+            body: r#"{"results":[{"rows":[[6,16]],"col_names":["length","length"]}]}"#,
+        },
+        // Too many parameters (OK)
+        TestCase {
+            query: "select $1 as col",
+            params: Some(vec![("1", "int"), ("2", "int8")]),
```

Comment:
Why is this ok?

Reply:
Ah--do you have any reference where unused parameters should generate errors? It's trivial to implement, but I couldn't find anything suggesting they should, but might very well be mistaken.

Reply:
```
send
Parse {"query": "SELECT $1 + 1"}
Bind {"values": ["1", "2"]}
Sync
----

until
ReadyForQuery
----
ParseComplete
ErrorResponse {"fields":[{"typ":"S","value":"ERROR"},{"typ":"C","value":"08P01"},{"typ":"M","value":"bind message supplies 2 parameters, but prepared statement \"\" requires 1"}]}
ReadyForQuery {"status":"I"}
```

Here's a pgtest that generates an error against postgres. We don't currently test this but we should.